pkg/workflows/management/create_cluster.go

```diff
@@ -0,0 +1,379 @@
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
+
+	return &installEksaComponentsTask{}
+}
+
+func (s *createBootStrapClusterTask) Name() string {
+	return "bootstrap-cluster-init"
+}
+
+func (s *createBootStrapClusterTask) Restore(ctx context.Context, commandContext *task.CommandContext, completedTask *task.CompletedTask) (task.Task, error) {
+	return nil, nil
+}
+
+func (s *createBootStrapClusterTask) Checkpoint() *task.CompletedTask {
+	return nil
+}
+
+func (s *installEksaComponentsTask) Run(ctx context.Context, commandContext *task.CommandContext) task.Task {
+	if !commandContext.BootstrapCluster.ExistingManagement {
+		logger.Info("Installing EKS-A custom components (CRD and controller) on workload cluster")
+		err := commandContext.ClusterManager.InstallCustomComponents(ctx, commandContext.ClusterSpec, commandContext.BootstrapCluster, commandContext.Provider)
+		if err != nil {
+			commandContext.SetError(err)
+			return &workflows.CollectDiagnosticsTask{}
+		}
+		logger.Info("Installing EKS-D components on workload cluster")
+		err = commandContext.EksdInstaller.InstallEksdCRDs(ctx, commandContext.ClusterSpec, commandContext.BootstrapCluster)
+		if err != nil {
+			commandContext.SetError(err)
+			return &workflows.CollectDiagnosticsTask{}
+		}
+	}
+
+	logger.Info("Creating EKS-A CRDs instances on workload cluster")
+	err := commandContext.ClusterManager.CreateEKSAReleaseBundle(ctx, commandContext.BootstrapCluster, commandContext.ClusterSpec)
+	if err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectDiagnosticsTask{}
+	}
+	err = commandContext.EksdInstaller.InstallEksdManifest(ctx, commandContext.ClusterSpec, commandContext.BootstrapCluster)
+	if err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectDiagnosticsTask{}
+	}
+
+	logger.Info("Applying cluster spec to bootstrap cluster")
+	if err := commandContext.ClusterUpgrader.Run(ctx, commandContext.ClusterSpec, *commandContext.BootstrapCluster); err != nil {
+		commandContext.SetError(err)
+		return &workflows.CollectMgmtClusterDiagnosticsTask{}
+	}
+
+	return &createWorkloadClusterTask{}
+}
+
+func (s *installEksaComponentsTask) Name() string {
+	return "eksa-components-install"
+}
+
+func (s *installEksaComponentsTask) Restore(ctx context.Context, commandContext *task.CommandContext, completedTask *task.CompletedTask) (task.Task, error) {
+	return nil, nil
+}
+
+func (s *installEksaComponentsTask) Checkpoint() *task.CompletedTask {
+	return nil
+}
+
+// createWorkloadClusterTask implementation
+
+func (s *createWorkloadClusterTask) Run(ctx context.Context, commandContext *task.CommandContext) task.Task {
+	logger.Info("Creating new workload cluster")
+	workloadCluster, err := commandContext.ClusterManager.GetWorkloadCluster(ctx, commandContext.BootstrapCluster, commandContext.ClusterSpec, commandContext.Provider)
```

Comment:
i have my doubts about this method

first i would prefer if we don't implement it in cluster manager
that's a super overloaded entity and tbh, not a good abstraction (I am the one to be blamed there)
It can still be in the clustermanager package if that makes it a bit easier for you, but ideally it should not depend on the clustermanager object.

Second, I wonder if we can just group it with the applier to create a higher level abstraction. Like a `ClusterCreator` or `CreateCluster` (horrible name, feel free to choose a better one). This entity would take the cluster spec and the kubeconfig for the "management cluster" (bootstrap in this occasion) and return the `types.Cluster` object for the newly created workload cluster. It can use the applier internally. It's a similar concept to what you can find here https://github.com/aws/eks-anywhere/blob/main/pkg/workflow/task/workload/create.go (minus the part that ties to the `Task` interface. Alternatively you could have a `CreateCluster` with two methods, `Create` (or `CreateSync`) and `WriteKubeconfig`.

I believe this would also be reusable for the workload cluster flow.

Reply:
https://github.com/orgs/aws/projects/162/views/1?pane=issue&itemId=46266515